onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlSTFT.h

```diff
@@ -0,0 +1,610 @@
+            this->batchSize = dims[0];
+            this->signalSize = dims[1];
+
+            if (rank == 3)
+            {
+                ML_CHECK_VALID_ARGUMENT(dims[2] == 1, "DML STFT only accepts real-valued input.");
+            }
+
+            MLOperatorEdgeDescription edgeDesc;
+            ORT_THROW_IF_FAILED(context->GetInputEdgeDescription(0, &edgeDesc));
+
+            assert(edgeDesc.edgeType == MLOperatorEdgeType::Tensor);
+            this->dataType = Dml::GetDmlDataTypeFromMlDataType(edgeDesc.tensorDataType);
+        }
+
+        // input 1: frame_step (required; constant; scalar)
+        {
+            ComPtr<IMLOperatorTensor> tensorInterface;
+            ORT_THROW_IF_FAILED(context->GetConstantInputTensor(DmlSTFTKernelInputIndex::FrameStep, &tensorInterface));
+
+            MLOperatorTensor tensor(tensorInterface.Get());
+            this->frameStep = onnxruntime::narrow<uint32_t>(OperatorHelper::ReadScalarTensorCastToInt64(tensor));
+        }
+
+        // input 2: window (optional; tensor)
+        if (context->IsInputValid(DmlSTFTKernelInputIndex::Window))
+        {
+            uint32_t rank;
+            ORT_THROW_IF_FAILED(shapeInterface->GetInputTensorDimensionCount(DmlSTFTKernelInputIndex::Window, &rank));
+            ML_CHECK_VALID_ARGUMENT(rank == 1, "Window shape must be 1D.");
+
+            ORT_THROW_IF_FAILED(shapeInterface->GetInputTensorShape(DmlSTFTKernelInputIndex::Window, rank, &this->frameSize));
+
+            this->hasWindowTensor = true;
+        }
+
+        // input 3: frame_length (optional; constant; scalar)
+        if (context->IsInputValid(DmlSTFTKernelInputIndex::FrameLength))
+        {
+            ComPtr<IMLOperatorTensor> tensorInterface;
+            ORT_THROW_IF_FAILED(context->GetConstantInputTensor(DmlSTFTKernelInputIndex::FrameLength, &tensorInterface));
+
+            MLOperatorTensor tensor(tensorInterface.Get());
+            uint32_t frameLength = onnxruntime::narrow<uint32_t>(OperatorHelper::ReadScalarTensorCastToInt64(tensor));
+
+            ML_CHECK_VALID_ARGUMENT(
+                this->frameSize == 0 || this->frameSize == frameLength,
+                "The window size and frame_length must be equal, if both are provided."
+            );
+
+            this->frameSize = frameLength;
+        }
+
+        ML_CHECK_VALID_ARGUMENT(this->frameSize > 0, "Either the window or frame_length must be set.");
+
+        // This limitation is a result of GpuDFTOperator in DmlDFT.h only implementing the stockham path.
+        // Remove this check if GpuDFTOperator is updated to support BluesteinZChirp.
+        ML_CHECK_VALID_ARGUMENT(
+            DFTHelpers::IsPowerOfTwo(this->frameSize),
+            "DML STFT only supports power-of-two window/frame sizes."
+        );
+
+        this->frameCount = (this->signalSize - this->frameSize) / this->frameStep + 1;
```

Comment:
>this->signalSize - this->frameSize [](http://example.com/codeflow?start=28&length=34)

Can this underflow? (if `frameSize` > `signalSize`)
Should we add a check that frameStep > 0? Don't want to divide by zero on rogue model file input.

Reply:
Good suggestion.